File: intan_pariwara/intan_pariwara/custom/pick_list.py
# Copyright (c) 2025, DAS and Contributors
# License: GNU General Public License v3. See license.txt

import frappe
from frappe.model.mapper import get_mapped_doc
from frappe.utils import flt, now_datetime

from intan_pariwara.intan_pariwara.doctype.packing_list.packing_list import get_items

def set_branch(self, method=None):
	if self.purpose == "Material Transfer":
		self.branch = frappe.get_value("Material Request", self.material_request, "custom_branch")
	elif self.purpose == "Delivery":
		self.branch = frappe.get_value("Sales Order", self.locations[0].sales_order, "branch")

@frappe.whitelist()
def make_packing_list(source_name, target_doc=None):
	def set_missing_values(source, target):
		target.run_method("set_missing_values")
		if target.purpose == "Siplah Titipan":
			target.reference = source.pre_order

		for d in get_items(source.name, source.purpose):
			if d.qty_per_koli:
				# Hitung kemasan utuh dan sisa
				full_packs, _ = divmod(d.remaining_qty, d.qty_per_koli)
				# Tambahkan kemasan utuh ke package
				target.extend("items", [{**d, "qty": d.qty_per_koli} for _ in range(int(full_packs))])
			else:
				target.append("items", d)

		target.run_method("calculate_total")

	doclist = get_mapped_doc(
		"Pick List",
		source_name,
		{
			"Pick List": {
				"doctype": "Packing List",
				"field_map": {
					"name": "pick_list",
					"custom_petugas_picking": "petugas_packing",
					"custom_nama_petugas": "nama_petugas",
					"letter_head": "letter_head"
				},
				"validation": {"docstatus": ["=", 1]},
			}
		},
		target_doc,
		set_missing_values,
	)

	return doclist

@frappe.whitelist()
def print_counter(doctype,docname, counter):
	frappe.db.set_value(doctype, docname, {
		'custom_print_counter': counter,
		'custom_last_print': now_datetime()
	})

	frappe.db.commit()
	doc = frappe.get_doc(doctype, docname)
	return doc

File: intan_pariwara/intan_pariwara/custom/pick_list.js
// Copyright (c) 2025, DAS and contributors
// For license information, please see license.txt
frappe.ui.form.off("Pick List", "material_request")
frappe.ui.form.off("Pick List", "add_get_items_button")
frappe.ui.form.on("Pick List", {
    refresh: function (frm) {
        if (frm.doc.docstatus == 1) {
			if (
				frm.doc.status !== "Completed" &&
				flt(frm.doc.per_packing) < 100
			) {
				if (frappe.model.can_create("Packing List")) {
					frm.add_custom_button(
						__("Packing List"),
						() => {
							frappe.model.open_mapped_doc({
								method: "intan_pariwara.intan_pariwara.custom.pick_list.make_packing_list",
								frm: frm,
								freeze: true,
								freeze_message: __("Creating Packing List ..."),
							});
						},
						__("Create")
					);
				}
			}
		}

		frm.remove_custom_button(__("Delivery Note"), __("Create"));
		frm.remove_custom_button(__("Stock Entry"),__("Create"));
    },
	material_request: (frm) => {
		frm.clear_table("locations")
		
		erpnext.utils.map_current_doc({
			method: "erpnext.stock.doctype.material_request.material_request.create_pick_list",
			target: frm,
			source_name: frm.doc.material_request,
		});
	},
	add_get_items_button: (frm) => {
		let purpose = frm.doc.purpose;
		if (purpose != "Delivery" || frm.doc.docstatus !== 0) {
			frm.remove_custom_button(__("Get Items"))
			return
		};
		let get_query_filters = {
			docstatus: 1,
			per_delivered: ["<", 100],
			status: ["!=", ""],
			customer: frm.doc.customer,
		};
		frm.get_items_btn = frm.add_custom_button(__("Get Items"), () => {
			erpnext.utils.map_current_doc({
				method: "erpnext.selling.doctype.sales_order.sales_order.create_pick_list",
				source_doctype: "Sales Order",
				target: frm,
				setters: {
					company: frm.doc.company,
					customer: frm.doc.customer,
				},
				date_field: "transaction_date",
				get_query_filters: get_query_filters,
			});
		});
	}
})